tests/python/model/test_ratex_fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,134 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Test ZeRO-1 implementation using FSDP interface."""
+import pytest
+
+import torch
+import torch.nn as nn
+from torchvision import datasets, transforms
+
+from raf.testing import get_dist_comm_info
+from raf import distributed as dist
+
+import ratex
+import ratex.lazy_tensor_core.core.lazy_model as lm
+from ratex.core.distributed.ratex_fully_sharded_data_parallel import RatexFullyShardedDataParallel
+from ratex.optimizer import SGD, Adam
+
+import numpy as np
+
+
+class SingleLayerLogistics(nn.Module):
+    def __init__(self, input_shape=28, num_classes=10):
+        super(SingleLayerLogistics, self).__init__()
+        self.log_softmax = nn.LogSoftmax(dim=-1)
+        self.linear = nn.Linear(784, num_classes)
+
+    def forward(self, x):
+        out = torch.flatten(x, 1)
+        out = self.linear(out)
+        out = torch.relu(out)
+        out = self.log_softmax(out)
+        return out
+
+
+def train(
+    device,
+    model,
+    optimizer,
+    optimizer_config,
+    image_datasets,
+    fsdp=False,
+    num_epochs=10,
+    dtype=torch.float32,
+):
+    dataloaders = {
+        x: torch.utils.data.DataLoader(
+            image_datasets[x], batch_size=1, shuffle=False, num_workers=1
+        )
+        for x in ["train", "val"]
+    }
+    dataset_sizes = {x: len(image_datasets[x]) for x in ["train", "val"]}
+    model = model.to(device, dtype=dtype)
+    if fsdp:
+        model = RatexFullyShardedDataParallel(model, optimizer, optimizer_config)
+        optimizer = model
+        model.train()
+    else:
+        model.train()
+        optimizer = optimizer(model.parameters(), **optimizer_config)
+    model = ratex.jit.script(model)
+
+    for epoch in range(num_epochs):
+        running_loss = 0.0
+
+        # Iterate over data.
+        for inputs, labels in dataloaders["train"]:
+            inputs = inputs.to(device)
+            inputs.requires_grad = True
+            labels_one_hot = torch.tensor(np.eye(10, dtype=np.float32)[labels])
+            labels_one_hot = labels_one_hot.to(device)  # One-hot
+            optimizer.zero_grad()
+            outputs = model(inputs)
+            loss = -torch.sum(outputs * labels_one_hot) / inputs.size(0)
+            loss.backward()
+            optimizer.step()
+            lm.mark_step()
+            running_loss += loss.item() * inputs.size(0)
+
+        epoch_loss = running_loss / dataset_sizes["train"]
+    return epoch_loss
+
+
+@pytest.mark.parametrize(
+    "optimizer", [(SGD, {"lr": 0.001, "momentum": 0.1}, 1), (Adam, {"lr": 0.001}, 2)]
+)
+def test_ratex_fully_sharded_data_parallelism_zero1(optimizer, tolerance=1e-10, seed=0):
+    torch.manual_seed(seed)
```

Comment:
You can use `@with_seed` with `train`, so that every call to `train` should have the same random seed and you can remove all `manual_seed` in this function.

An alternative is making dataloader in this function and pass to `train`, since the traverse order is identical for every epoch whatever for the same or different models.

Reply:
@with_seed does not seem to work for me. Perhaps it is because it does not say the pytorch seed? Do you think I should try to make my own decorator? Passing in data loader to train did not seem to work either without manually resetting the seed.

Reply:
Would that work if we put `torch.manual_seed` into `with_seed`? We can do that if so; otherwise we could just keep the current implementation.

Reply:
I believe that would work. I will give it a try, I just was not sure if it was appropriate to change the current implementation.

Reply:
I could not get it to work with with_seed for some reason. I am not sure if I am doing something incorrectly. I did however change the code so that torch.manual_seed is set and model is initialized in train, so that the code fragment is not duplicated